pkg/crypto/rings/client.go

```diff
@@ -0,0 +1,207 @@
+
+	"github.com/pokt-network/poktroll/pkg/client"
+	"github.com/pokt-network/poktroll/pkg/crypto"
+	pubkeyclient "github.com/pokt-network/poktroll/pkg/crypto/pubkey_client"
+	"github.com/pokt-network/poktroll/pkg/polylog"
+	"github.com/pokt-network/poktroll/x/service/types"
+)
+
+var _ crypto.RingClient = (*ringClient)(nil)
+
+type ringClient struct {
+	// logger is the logger for the ring cache.
+	logger polylog.Logger
+
+	// applicationQuerier is the querier for the application module, and is
+	// used to get the addresses of the gateways an application is delegated to.
+	applicationQuerier client.ApplicationQueryClient
+
+	// pubKeyClient
+	pubKeyClient crypto.PubKeyClient
+}
+
+// NewRingClient returns a new ring client constructed from the given dependencies.
+// It returns an error if the required dependencies are not supplied.
+//
+// Required dependencies:
+// - polylog.Logger
+// - client.ApplicationQueryClient
+// - client.AccountQueryClient
+func NewRingClient(deps depinject.Config) (_ crypto.RingClient, err error) {
+	rc := new(ringClient)
+	rc.pubKeyClient, err = pubkeyclient.NewPubKeyClient(deps)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := depinject.Inject(
+		deps,
+		&rc.logger,
+		&rc.applicationQuerier,
+	); err != nil {
+		return nil, err
+	}
+
+	return rc, nil
+}
+
+// GetRingForAddress returns the ring for the address provided. The ring is created by
+// querying for the application address and delegated gateways' account public keys and
+// converting them to their secp256k1 curve points.
+func (rc *ringClient) GetRingForAddress(
+	ctx context.Context,
+	appAddress string,
+) (*ring.Ring, error) {
+	pubKeys, err := rc.getDelegatedPubKeysForAddress(ctx, appAddress)
+	if err != nil {
+		return nil, err
+	}
+	// Cache the ring's points for future use
+	rc.logger.Debug().
+		Str("app_address", appAddress).
+		Msg("updating ring ringsByAddr for app")
+
+	// Get the points on the secp256k1 curve for the public keys in the ring.
+	points, err := pointsFromPublicKeys(pubKeys...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return the ring the constructed from the public key points on the secp256k1 curve.
+	return newRingFromPoints(points)
+}
+
+// VerifyRelayRequestSignature verifies the relay request signature against the
+// ring for the application address in the relay request.
+func (rc *ringClient) VerifyRelayRequestSignature(
+	ctx context.Context,
+	relayRequest *types.RelayRequest,
+) error {
+	rc.logger.Debug().
+		Fields(map[string]any{
+			"session_id":          relayRequest.Meta.SessionHeader.SessionId,
+			"application_address": relayRequest.Meta.SessionHeader.ApplicationAddress,
+			"service_id":          relayRequest.Meta.SessionHeader.Service.Id,
+		}).
+		Msg("verifying relay request signature")
+
+	// Extract the relay request's ring signature
+	if relayRequest.Meta == nil {
+		return ErrRingClientEmptyRelayRequestSignature.Wrapf(
+			"request payload: %s", relayRequest.Payload,
```

Comment:
`relayRequest.Payload` is a byte slice here. If you're sure about logging it, I would suggest formatting it as hex:

```suggestion
			"request payload: %x", relayRequest.Payload,
```

Reply:
The error used here is wrong. I'll change this to:
```go
if relayRequest.GetMeta() == nil {
  return ErrRingClientInvalidRelayRequest.Wrap("missing meta from relay request")
}
```

As there's not much useful information to provide about a nil field.